### PR TITLE
bf: ZENKO 1512 metastore circular json

### PR DESCRIPTION
**Pull request by dora-korpar**
Implements a serializer function for JSON.stringify to process circular JSON, based on this npm package: https://github.com/moll/json-stringify-safe/blob/master/stringify.js

### CODE REVIEW

**Review by rahulreddy — CHANGES_REQUESTED** [review]
It's good to do attribution for code that is copied over. You can use this https://meta.stackoverflow.com/questions/253618/if-i-use-so-code-on-my-website-how-should-i-give-attribution for reference.

**Comment by dora-korpar** [comment]
@rahulreddy I added a comment referencing the package with a link to the original

File: lib/storage/utils.js [hunk 1]
```diff
@@ -61,3 +61,39 @@ module.exports.setDirSyncFlag = function setDirSyncFlag(path, logger) {
         loggedWarning = true;
     }
 };
+
+// custom serializer function for JSON stringify to avoid circular JSON error
+// this code is a modified version of json-stringify-safe npm package found at
+// https://github.com/moll/json-stringify-safe/blob/master/stringify.js
+module.exports.serializeCircJSON = function serializeCircJSON() {
```
Comment (rahulreddy) [inline lib/storage/utils.js]: One more thing I noticed is that you have multiple module.exports floating in the file. This should have been caught by the linter but I guess not.
```js
function foo() {
 ........
}

function bar() {
.......
}
module.exports = {
foo, bar
}
```
This is the pattern we should be using for exports.

File: tests/unit/storage/utils.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+const circularJSON = {
+    key1: 'foo',
+    key2: 1,
+    key3: { key4: 5 },
+};
+circularJSON.key5 = circularJSON;
+
+describe('SerializeCircJSON', () => {
+    it('should stringify circular JSON', () => {
```
Comment (rahulreddy) [inline tests/unit/storage/utils.js:11]: Can you add a test stringifying a regular json and a test for invalid json?

**Review by rahulreddy — CHANGES_REQUESTED** [review]
Sorry I have two more comments

File: tests/unit/storage/utils.js (hunk 2 above)
Comment (dora-korpar) [reply tests/unit/storage/utils.js:11]: Added

File: lib/storage/utils.js (hunk 1 above)
Comment (dora-korpar) [reply lib/storage/utils.js]: Fixed

File: tests/unit/storage/utils.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+const assert = require('assert');
+const { serializeCircJSON } = require('../../../lib/storage/utils');
+
+const jsonObj = {
+    key1: 'foo',
+    key2: 1,
+    key3: { key4: 5 },
+};
+
+describe.only('SerializeCircJSON', () => {
```
Comment (philipyoo) [inline tests/unit/storage/utils.js]: ~.only~

**Review by philipyoo — APPROVED** [review]
what sorcery is this? 👀

File: tests/unit/storage/utils.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            assert.deepStrictEqual(parsedJSON.key3, { key4: 5 });
+        } catch (e) {
+            assert.ifError(e);
+        }
+    });
+
+    it('should fail to stringify invalid JSON', () => {
+        try {
+            //eslint-disable-next-line
+            const invalidJSON = { in$valid: value };
```
Comment (jonathan-gramain) [inline tests/unit/storage/utils.js]: Why is this object invalid or non convertible to JSON?

File: tests/unit/storage/utils.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
+const assert = require('assert');
+const { serializeCircJSON } = require('../../../lib/storage/utils');
+
+const jsonObj = {
```
Comment (jonathan-gramain) [inline tests/unit/storage/utils.js]: No big deal, but the naming can be confusing as this is a javascript object, not a JSON object.

**Review by jonathan-gramain — COMMENTED** [review]
I'm lacking context on why we have to handle circular references in the first place, as we should deal with object metadata which I would think should be simple flat JS objects. (anyway it's on mem backend so rather low risk but just wondering).

File: tests/unit/storage/utils.js (hunk 4 above)
Comment (dora-korpar) [reply tests/unit/storage/utils.js]: It's invalid because 'value' isn't defined anywhere

**Comment by dora-korpar** [comment]
As to the context, I wasn't able to actually reproduce the error in any testing, and the stack traces from the crashed Cloudservers weren't very forthcoming. It was decided it would be better to just catch the circular json than spend time trying to find the cause since it is just in the mem backend.

**Review by rahulreddy — APPROVED** [review]

**Comment by dora-korpar** [comment]
@bert-e approve

**Pull request merged**